File: apps/patter-voice-sdk/libraries/typescript/src/tunnel.ts
/**
 * Built-in tunnel support via cloudflared.
 *
 * Spawns a Cloudflare Quick Tunnel that exposes a local port to the internet.
 * Zero account required — uses Cloudflare's free trycloudflare.com service.
 *
 * Install: npm install cloudflared
 */

import { getLogger } from './logger';

const log = getLogger();

/** Handle returned by `startTunnel` exposing the public hostname and a stopper. */
export interface TunnelHandle {
  /** Public hostname (no protocol), e.g. "random-name.trycloudflare.com" */
  hostname: string;
  /** Stop the tunnel process */
  stop: () => void;
}

/**
 * Minimal structural type for the subset of the ``cloudflared`` package we use.
 * The package ships its own d.ts but is an optional peer dep, so we declare
 * the shape locally to keep the import dynamic and the build green even when
 * the package is not installed.
 */
interface TunnelInstance {
  on(event: string, cb: (data: string) => void): void;
  stop(): void;
}

interface CloudflaredModule {
  /** cloudflared >=0.7: returns an EventEmitter-shaped object. */
  Tunnel?: {
    quick?: (url: string) => TunnelInstance;
  };
  /** cloudflared 0.5: returns ``{ url, stop }`` (or an EventEmitter). */
  tunnel?: (args: Record<string, string>) => TunnelInstance & {
    url?: Promise<string>;
  };
}

/**
 * Start a cloudflared quick tunnel pointing to the given local port.
 *
 * @param port - Local port to tunnel to
 * @param timeoutMs - How long to wait for the tunnel URL (default 30s)
 * @returns A handle with the public hostname and a stop function
 */
export async function startTunnel(port: number, timeoutMs = 30_000): Promise<TunnelHandle> {
  let tunnelMod: CloudflaredModule;
  try {
    tunnelMod = (await import('cloudflared' as string)) as CloudflaredModule;
  } catch {
    throw new Error(
      'Built-in tunnel requires the "cloudflared" package. Install it with:\n\n' +
      '  npm install cloudflared\n\n' +
      'Or provide your own webhookUrl instead of using tunnel: true.'
    );
  }

  log.info('Starting tunnel to localhost:%d ...', port);

  // cloudflared@0.7+ exposes Tunnel.quick() for quick tunnels.
  // cloudflared@0.5 used tunnel({ '--url': ... }) returning { url: Promise }.
  // We support both APIs for backward compatibility.

  const TunnelClass = tunnelMod.Tunnel;
  const hasQuick = TunnelClass && typeof TunnelClass.quick === 'function';

  let instance: TunnelInstance;

  if (hasQuick && TunnelClass?.quick) {
    // New API (cloudflared 0.7+): Tunnel.quick(url) returns EventEmitter
    instance = TunnelClass.quick(`http://localhost:${port}`);
  } else {
    // Old API (cloudflared 0.5): tunnel({ '--url': ... }) returns { url: Promise, stop }
    if (!tunnelMod.tunnel) {
      throw new Error(
        'Built-in tunnel: installed "cloudflared" package exposes neither ' +
          '`Tunnel.quick` nor `tunnel({ "--url" })`. Upgrade with `npm install cloudflared@latest`.',
      );
    }
    const result = tunnelMod.tunnel({ '--url': `http://localhost:${port}` });
    if (result.url && typeof result.url.then === 'function') {
      // Old API with Promise-based URL
      const tunnelUrl: string = await Promise.race([
        result.url as Promise<string>,
        new Promise<never>((_, reject) =>
          setTimeout(() => reject(new Error(
            `Tunnel failed to start within ${timeoutMs / 1000}s. ` +
            'Check your internet connection or provide webhookUrl manually.'
          )), timeoutMs)
        ),
      ]);

      const hostname = tunnelUrl.replace(/^https?:\/\//, '').replace(/\/$/, '');
      log.info('Tunnel ready: https://%s', hostname);
      return { hostname, stop: () => { log.info('Stopping tunnel...'); result.stop(); } };
    }
    // If result has .on (EventEmitter pattern), fall through to event-based handling
    instance = result;
  }

  // Event-based URL resolution (cloudflared 0.7+)
  const tunnelUrl: string = await Promise.race([
    new Promise<string>((resolve) => {
      instance.on('url', (url: string) => resolve(url));
    }),
    new Promise<never>((_, reject) =>
      setTimeout(() => reject(new Error(
        `Tunnel failed to start within ${timeoutMs / 1000}s. ` +
        'Check your internet connection or provide webhookUrl manually.'
      )), timeoutMs)
    ),
  ]);

  const hostname = tunnelUrl.replace(/^https?:\/\//, '').replace(/\/$/, '');
  log.info('Tunnel ready: https://%s', hostname);

  return {
    hostname,
    stop: () => {
      log.info('Stopping tunnel...');
      instance.stop();
    },
  };
}

File: apps/patter-voice-sdk/libraries/python/getpatter/tunnel.py
"""Built-in tunnel support via cloudflared.

Spawns a Cloudflare Quick Tunnel that exposes a local port to the internet.
Zero account required — uses Cloudflare's free trycloudflare.com service.

Requires the ``cloudflared`` binary to be installed and on PATH.
Install: https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/downloads/
"""

from __future__ import annotations

import asyncio
import atexit
import logging
import re
import shutil
from collections.abc import Callable
from dataclasses import dataclass

logger = logging.getLogger("getpatter")

_TUNNEL_URL_RE = re.compile(r"https://([a-zA-Z0-9._-]+\.trycloudflare\.com)")
_STARTUP_TIMEOUT = 30  # seconds


@dataclass(frozen=True)
class TunnelHandle:
    """Handle to a running cloudflared tunnel."""

    hostname: str
    """Public hostname (no protocol), e.g. 'random-name.trycloudflare.com'."""

    stop: Callable[[], None]
    """Stop the tunnel process."""


async def start_tunnel(port: int, timeout: float = _STARTUP_TIMEOUT) -> TunnelHandle:
    """Start a cloudflared quick tunnel pointing to the given local port.

    Args:
        port: Local port to tunnel to.
        timeout: How long to wait for the tunnel URL (default 30s).

    Returns:
        A ``TunnelHandle`` with the public hostname and a stop function.

    Raises:
        FileNotFoundError: If cloudflared binary is not installed.
        TimeoutError: If the tunnel doesn't produce a URL within ``timeout``.
    """
    binary = shutil.which("cloudflared")
    if binary is None:
        raise FileNotFoundError(
            "Built-in tunnel requires the 'cloudflared' binary. Install it:\n\n"
            "  brew install cloudflared          # macOS\n"
            "  sudo apt install cloudflared      # Debian/Ubuntu\n"
            "  https://developers.cloudflare.com/cloudflare-one/connections/connect-networks/downloads/\n\n"
            "Or provide your own webhook_url instead of using tunnel=True."
        )

    logger.info("Starting tunnel to localhost:%d ...", port)

    proc = await asyncio.create_subprocess_exec(
        binary,
        "tunnel",
        "--url",
        f"http://localhost:{port}",
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )

    hostname: str | None = None

    async def _read_stream(stream: asyncio.StreamReader) -> str | None:
        """Read lines from a stream looking for the tunnel URL."""
        while True:
            line_bytes = await stream.readline()
            if not line_bytes:
                break
            line = line_bytes.decode("utf-8", errors="replace")
            match = _TUNNEL_URL_RE.search(line)
            if match:
                return match.group(1)
        return None

    try:
        # cloudflared prints the URL to stderr
        tasks = []
        if proc.stderr:
            tasks.append(asyncio.create_task(_read_stream(proc.stderr)))
        if proc.stdout:
            tasks.append(asyncio.create_task(_read_stream(proc.stdout)))

        done, pending = await asyncio.wait(
            tasks,
            timeout=timeout,
            return_when=asyncio.FIRST_COMPLETED,
        )

        for task in done:
            result = task.result()
            if result:
                hostname = result
                break

        # Cancel remaining tasks and await them to prevent
        # "Task was destroyed but it is pending!" warnings in Python 3.12+
        for task in pending:
            task.cancel()
        if pending:
            await asyncio.gather(*pending, return_exceptions=True)

    except Exception:
        try:
            proc.terminate()
        except ProcessLookupError:
            pass  # cloudflared already exited (e.g. transient quick-tunnel failure)
        raise

    if hostname is None:
        try:
            proc.terminate()
        except ProcessLookupError:
            pass  # cloudflared already exited
        raise TimeoutError(
            f"Tunnel failed to start within {timeout}s. "
            "Check your internet connection or provide webhook_url manually."
        )

    def _stop() -> None:
        logger.info("Stopping tunnel...")
        try:
            proc.terminate()
        except ProcessLookupError:
            pass  # Already exited

    # Safety net: kill tunnel if process exits without calling stop
    atexit.register(_stop)

    logger.info("Tunnel ready: https://%s", hostname)

    return TunnelHandle(hostname=hostname, stop=_stop)
